### ISSUE TITLE
How you using ArcMarginProduct.py as loss

### ISSUE THREAD

**Issue by jaideep11061982** [wujiyang/Face_Pytorch#10]
hi how  are you using arcface as loss. I could only see Cross ENtropy loss in your implementation.

**Comment by wujiyang** [comment]
Arcface is implemented in margin/ArcMarginProduct.py

**Comment by jaideep11061982** [comment]
thanks..
what is thought process behind passing output of above to CrossEntropy()
 output = margin(raw_logits, label)
  total_loss = criterion(output, label)

**Comment by wujiyang** [comment]
Well I think you didn't understand the principle of margin-based algorithms (sphereface, cosface, arcface).
They all use Cross Entropy as loss function, the difference between them lies in the softmax operation, different algorithms adopt different margins for the final classification vector.

**Comment by jaideep11061982** [comment]
in Arcface margin  you are  doing random init of self.weights . 
In every batch run  how would self.weights get updated as they not part of your net like resnet,mobilnet etc.  so every time you use only random weight you got first time during init.

 ```
for data in trainloader:
            img, label = data[0].to(device), data[1].to(device)
            optimizer_ft.zero_grad()

            raw_logits = net(img)
            output = margin(raw_logits, label)
            total_loss = criterion(output, label)
            total_loss.backward()
```

**Comment by williamCorrea85** [comment]
If you want the arcface weights to be updated after each backpropagation, you have to add them to the initialisation of your optimizer. 
Sometime like optmizer = optim.Adam(list(model.parameters()) + list(arcface.parameters(), lr = your_learning_rate)
